R/MsBackendDataFrame-functions.R

```diff
@@ -238,3 +233,65 @@ MsBackendDataFrame <- function() {
             do.call(.SPECTRA_DATA_COLUMNS[column], args = list())
     } else stop("column '", column, "' not available")
 }
+
+#' @description
+#'
+#' Helper to be used in the filter functions to select the file/origin in
+#' which the filtering should be performed.
+#'
+#' @param object `MsBackend`
+#'
+#' @param dataStorage `character` or `integer` with either the names of the
+#'     `dataStorage` or their index (in `dataStorageNames(object)`) in which
+#'     the filtering should be performed.
+#'
+#' @param dataOrigin same as `dataStorage`, but for the `dataOrigin` spectra
+#'     variable.
+#'
+#' @return `logical` of length equal to the number of spectra in `object`.
+#'
+#' @noRd
+.sel_file <- function(object, dataStorage = integer(), dataOrigin = integer()) {
+    if (length(dataStorage)) {
+        if (!(is.numeric(dataStorage) | is.character(dataStorage)))
+            stop("'dataStorage' has to be either an integer with the index of",
+                 " the data storage, or its name")
+        if (is.numeric(dataStorage))
+            dataStorage <- dataStorageNames(object)[dataStorage]
+        dataStorage(object) %in% dataStorage
+    } else if (length(dataOrigin)) {
+        if (!(is.numeric(dataOrigin) | is.character(dataOrigin)))
+            stop("'dataOrigin' has to be either an integer with the index of",
+                 " the data origin, or its name")
+        if (is.numeric(dataOrigin))
+            dataOrigin <- unique(dataOrigin(object))[dataOrigin]
+        dataOrigin(object) %in% dataOrigin
+    } else rep(TRUE, length(object))
+}
+
+#' Helper function to combine backends that base on [MsBackendDataFrame()].
+#'
+#' @param objects `list` of `MsBackend` objects.
+#'
+#' @return [MsBackend()] object with combined content.
+#'
+#' @author Johannes Rainer
+#'
+#' @noRd
+.combine_backend_data_frame <- function(objects) {
+    if (length(objects) == 1)
+        return(objects[[1]])
+    if (!all(vapply(objects, class, character(1)) == class(objects[[1]])))
+        stop("Can only merge backends of the same type: ", class(objects[[1]]))
+    res <- new(class(objects[[1]]))
+    suppressWarnings(
+        res@spectraData <- .as_rle_spectra_data(do.call(
+            .rbind_fill, lapply(objects, function(z) z@spectraData)))
+    )
+    if (any(colnames(res@spectraData) == "mz"))
+        res@spectraData$mz[any(is.na(res@spectraData$mz))] <- list(numeric())
```

Comment:
Hell yes! You're right. This should be `is.na`. Probably it would have worked anyway because I'd assume that a `Spectra` either has m/z and intensity values for all spectra or for none - but it would have failed otherwise. Thanks @sgibb !!!

Reply:
Fixed now.